content/en/blog/_posts/2022-10-27-live-and-let-live-with-kluctl-and-ssa.md

```diff
@@ -0,0 +1,172 @@
+either respond with an OK response or with a CONFLICT response.
+
+In case of a CONFLICT response, the apiserver also sends a status object with
+details about the conflicts. Kluctl can then use these details to figure out
+which fields are in conflict and which actors (field managers) have taken
+ownership of the conflicted fields.
+
+Then, for each field, Kluctl will decide if the conflict should be ignored or
+if it should be force-applied. If any field needs to be force-applied, Kluctl
+will retry the apply operation with the ignored fields omitted and the `force`
+flag being set on the API call.
+
+In case a conflict is ignored, Kluctl will issue a warning to the user so that
+the user can properly react (or ignore it forever...).
+
+That's basically it. That is all that is required to leverage Server-side apply.
+Big thanks and thumbs-up to the Kubernetes developers who made this possible!
+
+## Deciding which conflicts to ignore and which to force-apply
+
+Kluctl has a few simple rules to figure out if a conflict should be ignored
+or force-applied.
+
+It first checks the field's actor (the field manager) against a list of known
+tools that are well known to be used to perform manual modifications. These
+are for example kubectl and k9s. Any modifications performed with these tools
+are considered "temporary" and will be overwritten by Kluctl.
+
+If the field manager is not known by Kluctl, it will check if a force-apply is
+requested for that field. A force-apply can be requested in different ways:
+
+1. By passing `--force-apply` to Kluctl. This will cause ALL fields to be force-applied on conflicts.
+2. By adding the [`kluctl.io/force-apply=true`](https://kluctl.io/docs/reference/deployments/annotations/all-resources/#kluctlioforce-apply) annotation to the object in question. This will cause all fields of that object to be force-applied on conflicts.
+3. By adding the [`kluctl.io/force-apply-field=my.json.path`](https://kluctl.io/docs/reference/deployments/annotations/all-resources/#kluctlioforce-apply-field) annotation to the object in question. This cause only fields matching the JSON path to be force-applied on conflicts.
+
+## Future work in regard to conflict resolution
+
+I plan to add more ways to control the behaviour of conflict resolution. For
+example:
+
+1. It will be possible to force-apply on field level via the CLI.
+2. It will be possible to force-apply on field level but only for specific field managers.
+3. It will be possible to modify the list of known actors for which fields are always force-applied.
+4. Exclude fields from force-apply even if the field manager is in the list of well known tools.
+5. If more is needed, you can [open an issue](https://github.com/kluctl/kluctl/issues).
+
+## DevOps vs. PipelineOps vs. GitOps vs. Controllers
+
+So how does Server-side apply in Kluctl lead to "live and let live"?
+
+It allows the co-existence of classical DevOps, PipelineOps, GitOps and
```

Comment:
I'm not sure that “classical DevOps” was a thing - the people who introduced me to DevOps over a decade ago were already using it in the way that people now call GitOps.

Reply:
For me "classical DevOps" is the early stage of "everything as code" and then "running automation scripts from your laptop and from your pipelines". Do you have an idea for a better name of this form of "DevOps"?

Reply:
You could write “API defined infrastructure” or “infrastructure as code”. How about a new wording eg:

“_Wherever you are on your infrastructure automated journey, Kluctl has a place for you.
From running deploys using a script on your PC, all the way to fully automated CI/CD with the
pipelines themselves defined in code, Kluctl aims to complement the workflow that's right for you._”

Feel free to reword that to align with your own writing style and authentic voice.